### ISSUE TITLE
[Feature] Improve the robustness of keywords processing

### ISSUE THREAD

**Issue by olivierobert** [thanakij/serp-web#3]
## Issue

Keywords are processed as per the following: 1) the user uploads CSV (Next.JS), 2) keywords are sent to `/api/keywords` (Next.JS), and 3) keywords are sent to the API application. Even though the API_URL is named WORKER_URL, it is not a queue system. There is also no adequate error handling (e.g., https://github.com/thanakij/serp-api/blob/main/src/main.js#L87), so if any error occurs, the processing of all keywords fails 💥 The core business logic is too fragile.

## Expected

- Keywords should be stored in the database with a flag to track the scraping status.
- A distinct background job to scrap the Google page is then scheduled.

The benefits are:

- Fast CSV upload request
- The scraping of each keyword is isolated, i.e., one could error while others could be successful.

**Comment by thanakij** [comment]
@olivierobert You are right that the keywords are uploaded (to _/api/keywords_ in **serp-web**) and then transformed as JSON data (an array of keywords) to **serp-api** repo.

I would like to emphasize that I purposefully didn't put the keyword **await** in [L16](https://github.com/thanakij/serp-web/blob/main/pages/api/keywords/index.js#L16) so the code will continue without waiting. In short, it is already fast. However, I totally agree that it can be improved by leveraging a queue system. This way, API can store the job and respond to the nextjs frontend immediately. As a result, we then can add the **await** keyword to ensure about the acknowledgement from API.

My current code is looping over keyword one by one ([L70](https://github.com/thanakij/serp-api/blob/main/src/main.js#L70)) so the keywords are processed independently. Next, not only there exists a retry logic in place ([L73](https://github.com/thanakij/serp-api/blob/main/src/main.js#L73)), but also a `try...catch` block ([L77](https://github.com/thanakij/serp-api/blob/main/src/main.js#L77)). Therefore, I would like to defend that it is not super fragile :P. Anyway, I totally agree that a **status** flag should be added. This way, the UX of the user is improved as the user can know immediately that the keywords are already enqueued and waiting to be processed. In addition, in case of any error during scraping, we can stamp the record as error so the user is aware of this as well (instead of totally skipping the record).

**Comment by thanakij** [comment]
In terms of implementation, after we insert the records into the database and retrieve the set of inserted IDs. We then can put them into the queue with one message per ID. Then a set of consumers on the same queue will receive a message and process for that one single ID (ie, one keyword). Logically, we can speed up the process by spawning more than 1 consumer. Also, we will need to experiment about this parameter to ensure that we won't get banned by Google due to the concurrent requests.

**Comment by olivierobert** [comment]
While the below code processes keywords one by one, the processing is sequential (one keyword after another one):

https://github.com/thanakij/serp-api/blob/82cf3ebe706c8cbfc30fcbabf07e81265c9b9020/src/main.js#L70-L89

Despite having the `try/catch,` any unexpected error would stop the loop. That is where a queue system or an async process that picks up background jobs would be more reliable. This was the code of the technical challenge, hence the focus on it in the code review 🔍
